tensorflow/python/ops/array_ops.py

```diff
@@ -1757,6 +1757,176 @@ def _DepthToSpaceShape(op):
       [input_shape[0], height, width, new_depth])]
 
 
+def one_hot(indices, depth, on_value=None, off_value=None,
+            axis=None, dtype=None, name=None):
```

Comment:
Okay, how about the following compromise: make the default `dtype=tf.float32`.  Once we always have a dtype, the whole case analysis becomes the two lines

    on_value = ops.convert_to_tensor(on_value, dtype, name="on_value")
    off_value = ops.convert_to_tensor(off_value, dtype, name="off_value")